exercises/qos/README.md

```diff
@@ -0,0 +1,166 @@
+`qos.p4`, which initially implements L3 forwarding. Your job (in the
+next step) will be to extend it to properly append set the `diffserv` bits.
+
+Before that, let's compile the incomplete `qos.p4` and bring up a
+network in Mininet to test its behavior.
+
+1. In your shell, run:
+   ```bash
+   make
+   ```
+   This will:
+   * compile `qos.p4`, and
+   * start a Mininet instance with three switches (`s1`, `s2`, `s3`) configured
+     in a triangle. There are 5 hosts. `h1` and `h11` are connected to `s1`.
+     `h2` and `h22` are connected to `s2` and `h3` is connected to `s3`.
+   * The hosts are assigned IPs of `10.0.1.1`, `10.0.2.2`, etc
+     (`10.0.<Switchid>.<hostID>`).
+   * The control plane programs the P4 tables in each switch based on
+     `sx-runtime.json`
+
+2. We want to send traffic from `h1` to `h2`. If we
+capture packets at `h2`, we should see the right diffserv value.
+
+![Setup](setup.png)
+
+3. You should now see a Mininet command prompt. Open two terminals
+for `h1` and `h2`, respectively:
+   ```bash
+   mininet> xterm h1 h2
+   ```
+4. In `h2`'s XTerm, start the server that captures packets:
+   ```bash
+   ./receive.py
+   ```
+5. In `h1`'s XTerm, send one packet per second to `h2` using send.py
+say for 30 seconds.
+   To send UDP:
+   ```bash
+   ./send.py --p=UDP --des=10.0.2.2 --m="P4 is cool" --dur=30
+   ```
+   To send TCP:
+   ```bash
+   ./send.py --p=TCP --des=10.0.2.2 --m="P4 is cool" --dur=30
+   ```
+   The message "P4 is cool" should be received in `h2`'s xterm,
+6. At `h2`, the `ipv4.tos` field (DiffServ+ECN) is always 1
+7. type `exit` to close each XTerm window
+
+Your job is to extend the code in `qos.p4` to implement the diffserv logic
+for setting the diffserv flag.
+
+## Step 2: Implement Diffserv
+
+The `qos.p4` file contains a skeleton P4 program with key pieces of
+logic replaced by `TODO` comments.  These should guide your
+implementation---replace each `TODO` with logic implementing the
+missing piece.
+
+First we have to change the ipv4_t header by splitting the TOS field
+into DiffServ and ECN fields.  Remember to update the checksum block
+accordingly.  Then, in the egress control block we must compare the
+protocol in IP header with IP protocols. Based on the traffic classes 
+and priority, the `diffserv` flag will be set.
+
+A complete `qos.p4` will contain the following components:
+
+1. Header type definitions for Ethernet (`ethernet_t`) and IPv4 (`ipv4_t`).
+2. Parsers for Ethernet, IPv4,
+3. An action to drop a packet, using `mark_to_drop()`.
+4. An action (called `ipv4_forward`), which will:
+	1. Set the egress port for the next hop.
+	2. Update the ethernet destination address with the address of
+           the next hop.
+	3. Update the ethernet source address with the address of the switch.
+	4. Decrement the TTL.
+5. An ingress control block that checks the protocols and sets the ipv4.diffserv.
+6. A deparser that selects the order in which fields inserted into the outgoing
```

Comment:
Maybe replace "in which fields inserted" with "in which headers are inserted" ?

Reply:
Done